pkg/spao/mac.go

```diff
@@ -123,93 +181,223 @@ func serializeAuthenticatedData(
 	if err != nil {
 		return 0, err
 	}
-	offset += s.Path.Len()
+	offset += s.Path.Len( /*@ ubuf @*/ )
 	return offset, nil
 }
 
-func zeroOutMutablePath(orig path.Path, buf []byte) error {
-	err := orig.SerializeTo(buf)
+// @ requires  orig != nil
+// @ requires  len(buf) >=  28 + 12 * scion.MaxHops + epic.MetadataLen
+// @ requires  acc(orig.Mem(ubuf), R9)
+// @ requires  (typeOf(orig) == type[*scion.Raw] ==>
+// @				orig.(*scion.Raw).GetBase(ubuf).WeaklyValid()) &&
+// @		   (typeOf(orig) == type[*scion.Decoded] ==>
+// @				orig.(*scion.Decoded).GetBase(ubuf).WeaklyValid()) &&
+// @		   (typeOf(orig) == type[*epic.Path] ==>
+// @				orig.(*epic.Path).GetBase(ubuf).WeaklyValid())
+// @ preserves sl.Bytes(buf, 0, len(buf))
+// @ preserves sl.Bytes(ubuf, 0, len(ubuf))
+// @ ensures   acc(orig.Mem(ubuf), R9)
+// @ ensures   retErr != nil ==> retErr.ErrorMem()
+// @ decreases
+func zeroOutMutablePath(orig path.Path, buf []byte /*@, ghost ubuf []byte @*/) (retErr error) {
+	err := orig.SerializeTo(buf /*@, ubuf @*/)
 	if err != nil {
 		return serrors.Wrap("serializing path for resetting fields", err)
 	}
 	switch p := orig.(type) {
 	case empty.Path:
 		return nil
 	case *scion.Raw:
+		// @ unfold acc(p.Mem(ubuf), R50)
+		// @ unfold acc(p.Base.Mem(), R50)
 		zeroOutWithBase(p.Base, buf)
+		// @ fold acc(p.Base.Mem(), R50)
+		// @ fold acc(p.Mem(ubuf), R50)
 		return nil
 	case *scion.Decoded:
+		// @ unfold acc(p.Mem(ubuf), R50)
+		// @ unfold acc(p.Base.Mem(), R50)
 		zeroOutWithBase(p.Base, buf)
+		// @ fold acc(p.Base.Mem(), R50)
+		// @ fold acc(p.Mem(ubuf), R50)
 		return nil
 	case *epic.Path:
+		// @ unfold acc(p.Mem(ubuf), R50)
+		// @ unfold acc(p.ScionPath.Mem(ubuf[epic.MetadataLen:]), R50)
+		// @ unfold acc(p.ScionPath.Base.Mem(), R50)
+		// @ sl.SplitRange_Bytes(buf, epic.MetadataLen, len(buf), writePerm)
 		zeroOutWithBase(p.ScionPath.Base, buf[epic.MetadataLen:])
+		// @ sl.CombineRange_Bytes(buf, epic.MetadataLen, len(buf), writePerm)
+		// @ fold acc(p.ScionPath.Base.Mem(), R50)
+		// @ fold acc(p.ScionPath.Mem(ubuf[epic.MetadataLen:]), R50)
+		// @ fold acc(p.Mem(ubuf), R50)
 		return nil
 	case *onehop.Path:
 		// Zero out IF.SegID
+		// @ unfold sl.Bytes(buf, 0, len(buf))
+		// @ assert forall j int :: { &buf[2:][j] } 0 <= 2 ==> &buf[2:][j] == &buf[2 + j]
 		binary.BigEndian.PutUint16(buf[2:], 0)
 		// Zero out HF.Flags&&Alerts
 		buf[8] = 0
 		// Zero out second HF
-		copy(buf[20:], []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0})
+		// @ assert forall j int :: { &buf[20:][j] } 0 <= 12 ==> &buf[20:][j] == &buf[20 + j]
+		copy(buf[20:], []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} /*@ , R20 @*/)
+		// @ fold sl.Bytes(buf, 0, len(buf))
 		return nil
 	default:
 		return serrors.New(fmt.Sprintf("unknown path type %T", orig))
 	}
 }
 
+// @ requires  len(buf) >= 28 + 12 * scion.MaxHops
+// @ preserves base.WeaklyValid()
```

Comment:
This should be a requires because it is a pure assertion: First, `base` is passed by value, not by reference. Snd, `base` in the postcondition still refers to the value stored in `base` in the beginning, as the value stored in it at the end is useless for the client of this function (`base` is, after all, a local variable) - we can discuss this point in a meeting.

At any rate, because of the reasons I mentioned, it is unnecessary to check this postcondition.
```suggestion
// @ requires base.WeaklyValid()
```

Reply:
Done.